### PR TITLE
fix(cd): the bug which can cd to a permission error place

### PR DESCRIPTION
**Pull request by Decodetalkers**
Log: use linux mode to judget wether can cd into the place

# Description

It is the fix for the issue that nushell can enter into a permission denied place

# Tests

Make sure you've done the following:

- [ ] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [ ] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [ ] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [ ] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [ ] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

**Comment by Decodetalkers** [comment]
I do not know the better solution, because it can simply define the variable of PWD, and enter to a place which is disallow...Maybe we need to set PWD cannot be defined?

**Comment by fdncred** [comment]
I'm not sure checking the permissions on a folder is a proper way to do it. The way you're doing it now it doesn't even seem like root could cd to some folders.

**Comment by Decodetalkers** [comment]
> I'm not sure checking the permissions on a folder is a proper way to do it. The way you're doing it now it doesn't even seem like root could cd to some folders.

Yes, I should also get who is the user now...So hao can I get the user..

**Comment by Decodetalkers** [comment]
Yes , I think I know how to realize that

**Comment by Decodetalkers** [comment]
> I'm not sure checking the permissions on a folder is a proper way to do it. The way you're doing it now it doesn't even seem like root could cd to some folders.

Now I check the user ,group id, can you review it?

**Comment by Decodetalkers** [comment]
Now , root, owner, other user, user in the same group, they all have different permission

**Comment by fdncred** [comment]
I'm not a big fan of the way this is checking user/group/etc. and it also isn't supported on Windows. Isn't there a cross platform way of doing this type of thing? I saw [this ](https://github.com/Freaky/faccess) but I'm not sure it's a good crate to use. Maybe there's something better?

**Comment by fdncred** [comment]
Another option is to look at tools like `exa` and `lsd` that are cross platform and see how they handle permissions.

**Comment by Decodetalkers** [comment]
> I'm not a big fan of the way this is checking user/group/etc. and it also isn't supported on Windows. Isn't there a cross platform way of doing this type of thing? I saw [this ](https://github.com/Freaky/faccess) but I'm not sure it's a good crate to use. Maybe there's something better?

I tried the crate, it cannot handle the problem..  I am also finding better solution

**Comment by Decodetalkers** [comment]
> Another option is to look at tools like `exa` and `lsd` that are cross platform and see how they handle permissions.

image
It seems that exa is the same? emmm

**Comment by Decodetalkers** [comment]
> Another option is to look at tools like `exa` and `lsd` that are cross platform and see how they handle permissions.

I find a new idea to solve the problem

File: crates/nu-command/src/filesystem/cd.rs [hunk 1]
```diff
@@ -169,11 +171,17 @@ impl Command for Cd {
         if let Some(oldpwd) = stack.get_env_var(engine_state, "PWD") {
             stack.add_env_var("OLDPWD".into(), oldpwd)
         }
-
         //FIXME: this only changes the current scope, but instead this environment variable
```
Comment (fdncred) [inline crates/nu-command/src/filesystem/cd.rs:174]: These comments go with the add_env_var("PWD") so it may be better to move them to where it is now

File: crates/nu-command/src/filesystem/cd.rs [hunk 2]
```diff
@@ -169,11 +171,17 @@ impl Command for Cd {
...
         //should probably be a block that loads the information from the state in the overlay
 
-        stack.add_env_var("PWD".into(), path_value);
+        // get a fake path ,is can get it , or not permission_denied , then can cd int
+        // the path
+        if permission_denied(PathBuf::from(format!("{}/enter", checkpath))) {
+            return Err(ShellError::IOError("Cannot into ".to_string()));
```
Comment (fdncred) [inline crates/nu-command/src/filesystem/cd.rs:180]: Maybe change this text from "Cannot into " to something more like "Permission denied changing to {} directory"

File: crates/nu-command/src/filesystem/cd.rs (hunk 1 above)
Comment (Decodetalkers) [reply crates/nu-command/src/filesystem/cd.rs]: > These comments go with the add_env_var("PWD") so it may be better to move them to where it is now

I have moved them now

File: crates/nu-command/src/filesystem/cd.rs [hunk 3]
```diff
@@ -170,10 +172,16 @@ impl Command for Cd {
             stack.add_env_var("OLDPWD".into(), oldpwd)
         }
 
-        //FIXME: this only changes the current scope, but instead this environment variable
-        //should probably be a block that loads the information from the state in the overlay
+        // get a fake path ,is can get it , or not permission_denied , then can cd int
+        // the path
+        if permission_denied(PathBuf::from(format!("{}/enter", checkpath))) {
```
Comment (fdncred) [inline crates/nu-command/src/filesystem/cd.rs:177]: So, you're using a fake file that may or may not exist. That's clever but I'm not sure it's the right thing to do.

**Review by fdncred — CHANGES_REQUESTED** [review]

File: crates/nu-command/src/filesystem/cd.rs (hunk 3 above)
Comment (fdncred) [reply crates/nu-command/src/filesystem/cd.rs:177]: it may be ok if the fake file was some system-generated-unique-filename that would never exist. we need more eyes on this to approve or disapprove.

Comment (fdncred) [reply crates/nu-command/src/filesystem/cd.rs:177]: @Decodetalkers can you change the word `enter` to some type of uniq system generated file name? I think, if we're going check for permission denied, we need to do it in such a way that there is really no possibility that the file exists and I think a unique name other than `enter` could do that.

Comment (Decodetalkers) [reply crates/nu-command/src/filesystem/cd.rs:177]: > @Decodetalkers can you change the word `enter` to some type of uniq system generated file name? I think, if we're going check for permission denied, we need to do it in such a way that there is really no possibility that the file exists and I think a unique name other than `enter` could do that.

Sorry, these days I am busy.. Now the problems is I don't know how to write the tests and what the uniq stystem generated file name should be...

Comment (Decodetalkers) [reply crates/nu-command/src/filesystem/cd.rs:177]: > @Decodetalkers can you change the word `enter` to some type of uniq system generated file name? I think, if we're going check for permission denied, we need to do it in such a way that there is really no possibility that the file exists and I think a unique name other than `enter` could do that.

Maybe "~"?, and will ~ available on macos and windows. However ~ is also available on linux, but always it should not be named as "~"

Comment (fdncred) [reply crates/nu-command/src/filesystem/cd.rs:177]: There is probably an idiomatic rust way to unique temporary file names. If not, there are probably crates to do this type of thing too.

**Comment by fdncred** [comment]
there doesn't seem to be much movement here, so closing for now.
